tests/sentry/integrations/utils/test_metrics.py

```diff
@@ -0,0 +1,186 @@
+                "integration_domain": "messaging",
+                "integration_name": "my_integration",
+                "interaction_type": "my_interaction",
+            },
+        )
+
+    @mock.patch("sentry.integrations.utils.metrics.logger")
+    @mock.patch("sentry.integrations.utils.metrics.metrics")
+    def test_recording_failure(self, mock_metrics, mock_logger):
+        metric_obj = self.TestLifecycleMetric()
+        with pytest.raises(ExampleException):
+            with metric_obj.capture() as lifecycle:
+                lifecycle.add_extra("extra", "value")
+                raise ExampleException
+
+        self._check_metrics_call_args(mock_metrics, "failure")
+        mock_logger.error.assert_called_once_with(
+            "integrations.slo.failure",
+            extra={
+                "extra": "value",
+                "integration_domain": "messaging",
+                "integration_name": "my_integration",
+                "interaction_type": "my_interaction",
+            },
+            exc_info=mock.ANY,
+        )
+
+    @mock.patch("sentry.integrations.utils.metrics.logger")
+    @mock.patch("sentry.integrations.utils.metrics.metrics")
+    def test_recording_explicit_failure_with_exception(self, mock_metrics, mock_logger):
+        metric_obj = self.TestLifecycleMetric()
+        with metric_obj.capture() as lifecycle:
+            try:
+                lifecycle.add_extra("extra", "value")
+                raise ExampleException
+            except ExampleException as exc:
+                lifecycle.record_failure(exc, extra={"even": "more"})
+
+        self._check_metrics_call_args(mock_metrics, "failure")
+        mock_logger.error.assert_called_once_with(
+            "integrations.slo.failure",
+            extra={
+                "extra": "value",
+                "even": "more",
+                "integration_domain": "messaging",
+                "integration_name": "my_integration",
+                "interaction_type": "my_interaction",
+            },
+            exc_info=mock.ANY,
+        )
+
+    @mock.patch("sentry.integrations.utils.metrics.logger")
+    @mock.patch("sentry.integrations.utils.metrics.metrics")
+    def test_recording_explicit_failure_with_str(self, mock_metrics, mock_logger):
+        metric_obj = self.TestLifecycleMetric()
+        with metric_obj.capture() as lifecycle:
+            lifecycle.add_extra("extra", "value")
+            lifecycle.record_failure("Integration went boom", extra={"even": "more"})
+
+        self._check_metrics_call_args(mock_metrics, "failure")
+        mock_logger.error.assert_called_once_with(
+            "integrations.slo.failure",
+            extra={
+                "outcome_reason": "Integration went boom",
+                "extra": "value",
+                "even": "more",
+                "integration_domain": "messaging",
+                "integration_name": "my_integration",
+                "interaction_type": "my_interaction",
```

Comment:
Most of these tests were just slightly modified, but because this also included a rename it considers this a new file 🤦🏼‍♂️ 

Notable changes:
- Tags are now included in the extra info (there was a bug that omitted it)
- failure and halt reasons are included as `outcome_reason` when strings are passed
- `exc_info` is conditionally passed if an exception type is passed as the outcome reason.